### PR TITLE
[stdlib] Update Array.subscript to use _modify

### PR DESCRIPTION
**Pull request by airspeedswift**
This switches `Array` from using `mutableAddressWithNativeOwner` to using a yielding `_modify`

### CODE REVIEW

**Comment by airspeedswift** [comment]
@swift-ci please smoke benchmark

**Comment by airspeedswift** [comment]
@swift-ci please test

**Comment by airspeedswift** [comment]
@swift-ci please smoke test compiler performance

**Comment by airspeedswift** [comment]
@swift-ci please test source compatibility

**Comment by airspeedswift** [comment]
I guess that'll teach me for skipping building the benchmarks 😞

File: stdlib/public/core/Array.swift [hunk 1]
```diff
@@ -646,31 +647,6 @@ extension Array {
     return _buffer.capacity
   }
 
-  /// - Precondition: The array has a native buffer.
-  @inlinable
-  @_semantics("array.owner")
```
Comment (slavapestov) [inline stdlib/public/core/Array.swift]: Can we remove optimizer code that checks for this semantics attribute also?

```
palomino:swift spestov$ git grep kGetArrayOwner
include/swift/SILOptimizer/Analysis/ArraySemantic.h:  kGetArrayOwner,
lib/SILOptimizer/Analysis/ArraySemantic.cpp:            .Case("array.owner", ArrayCallKind::kGetArrayOwner)
lib/SILOptimizer/Analysis/EscapeAnalysis.cpp:      case ArrayCallKind::kGetArrayOwner:
lib/SILOptimizer/LoopTransforms/COWArrayOpt.cpp:  case ArrayCallKind::kGetArrayOwner:
lib/SILOptimizer/LoopTransforms/COWArrayOpt.cpp:  case ArrayCallKind::kGetArrayOwner:
```

**Comment by slavapestov** [comment]
I did not expect the compile time wins. I wonder why?

**Comment by airspeedswift** [comment]
name | delta_pct
---: | ---: 
Frontend.NumInstructionsExecuted  | -21.56% :white_check_mark:

wat

**Comment by airspeedswift** [comment]
@swift-ci please test compiler performance

File: stdlib/public/core/Array.swift (hunk 1 above)
Comment (airspeedswift) [reply stdlib/public/core/Array.swift]: It's still there in `ContiguousArray` and `ArraySlice`, which are next up.

**Comment by airspeedswift** [comment]
@swift-ci please smoke benchmark

**Comment by airspeedswift** [comment]
@swift-ci please smoke test compiler performance

**Comment by gottesmm** [comment]
@airspeedswift I wonder if the memset issue is due to array hoisting failing. It could be that you are changing the pattern we are looking for so the semantic optimization for uniqueness check hoisting is not kicking in.

File: stdlib/public/core/Array.swift [hunk 2]
```diff
@@ -570,10 +570,11 @@ extension Array: RandomAccessCollection, MutableCollection {
         index, wasNativeTypeChecked: wasNativeTypeChecked,
         matchingSubscriptCheck: token)
     }
-    mutableAddressWithNativeOwner {
+    _modify {
       _makeMutableAndUnique() // makes the array native, too
       _checkSubscript_native(index)
-      return (_getElementAddress(index), _getOwner_native())
+      let address = _buffer.subscriptBaseAddress + index
```
[the remark this answers is not in the document]
Comment (airspeedswift) [reply stdlib/public/core/Array.swift]: Ok let me try putting that back. Seems a shame to have to have a symbol just for doing that.

[the remark this answers is not in the document]
Comment (eeckstein) [reply stdlib/public/core/Array.swift]: I already have a fix for this

**Comment by eeckstein** [comment]
https://github.com/apple/swift/pull/19181 should fix the crash and most of the regressions (except XorLoop, which needs more work)

**Comment by airspeedswift** [comment]
@swift-ci please smoke benchmark

**Comment by airspeedswift** [comment]
@swift-ci please test

**Comment by airspeedswift** [comment]
@eeckstein oddly getting a linux-only optimizer assertion now

**Comment by eeckstein** [comment]
@airspeedswift Can you disable the test for linux and file a bug?

**Comment by rjmccall** [comment]
Awesome!

**Pull request merged**